File: runtime/CSlib/symbol.cs
using System;

namespace bigloo {
   public sealed class symbol: obj {
      sealed class bucket {
	 public readonly symbol symb;
	 public bool marked = false;
	 public bucket next;
   
	 public bucket( symbol symb, bucket next ) {
	    this.symb = symb;
	    this.next = next;
	 }
      }

      public readonly byte[] pname;
      public Object cval;

      private static readonly bucket[] table = new bucket[4096];

      public symbol( Object cval, byte[] pname ) {
	 this.cval = cval;
	 this.pname = pname;
      }

      public static bool exists( byte[] name ) {
	 int hash_number = foreign.get_hash_power_number( name, 12 );
	 bucket _bucket = table[hash_number];

	 if (_bucket == null)
	    return false;

	 for (bucket run = _bucket ; (run != null) ; run = run.next)
	    if (foreign.bigloo_strcmp( run.symb.pname, name ))
	       return true;

	 return false;
      }

      public static symbol make_symbol( byte[] name ) {
	 lock( table ) {
	    int hash_number = foreign.get_hash_power_number( name, 12 );
	    bucket _bucket = table[hash_number];

	    if (_bucket == null) {
	       symbol _symbol = new symbol( nil._nil, name );

	       table[hash_number] = new bucket( _symbol, null );
	       return _symbol;
	    }

	    for (bucket run = _bucket ; (run != null) ; run = run.next)
	       if (foreign.bigloo_strcmp( run.symb.pname, name ))
		  return run.symb;

	    symbol result = new symbol( nil._nil, name );

	    table[hash_number] = new bucket( result, _bucket );

	    return result;
	 }
      }

      public override void write( output_port p ) {
	 p.write( pname );
      }
   }
}

File: runtime/Jlib/symbol.java
package bigloo;

import java.util.*;

public class symbol extends obj
{
   static final class bucket
   {
      public final symbol symb;
      public boolean marked = false;
      public bucket next;

      public bucket(final symbol symb, final bucket next) {
	 this.symb = symb;
	 this.next = next;
      }
   }

   public final byte[] string;
   public Object cval;

   public static int shift = 12;
   public static bucket[] table = new bucket[1 << shift];
   public static int cnt = 0;

   public symbol(final Object cval, final byte[] string) {
      this.cval = cval;
      this.string = string;
   }

   public static boolean exists(final byte[] name) {
      synchronized(table) {
	 final int hash_number = foreign.get_hash_power_number(name, shift);
	 final bucket bucket = table[hash_number];

	 if (bucket == null)
	    return false;

	 for (bucket run = bucket; (run != null); run = run.next)
	    if (foreign.bigloo_strcmp(run.symb.string, name))
	       return true;

	 return false;
      }
   }

   public static bucket[] resize_table(bucket[] old_table, int new_shift) {
      bucket[] new_table = new bucket[1 << new_shift];
      int old_size = old_table.length;
      int i;

      for (i = 0; i < old_size; i++) {
	 for (bucket run = old_table[i]; (run != null) ; run = run.next) {
	    symbol symb = run.symb;
	    int hash = foreign.get_hash_power_number(symb.string, new_shift);
	    new_table[hash] = new bucket(symb, new_table[hash]);
	 }
      }
      
      return new_table;
   }
   
   public static symbol make_symbol(final byte[] name) {
      synchronized(table) {
	 final int hash_number = foreign.get_hash_power_number(name, shift);
	 final bucket bucket = table[hash_number];

	 if (bucket == null) {
	    final symbol symbol = new symbol(nil.nil, name);

	    table[hash_number] = new bucket(symbol, null);
	    return symbol;
	 }

	 for (bucket run = bucket; (run != null) ; run = run.next)
	    if (foreign.bigloo_strcmp(run.symb.string, name))
	       return run.symb;

	 if (cnt++ > (1 << shift) * 4) {
	    shift++;
	    table = resize_table(table, shift);
	 } else {
	    final symbol result = new symbol(nil.nil, name);
	    table[hash_number] = new bucket(result, bucket);
	    return result;
	 }
      }
      // this point is reached when the table has been resized
      return make_symbol(name);
   }

   public void write(final output_port p) {
      p.write(string);
   }
}
